File: packages/csharp/src/Identifold/Storage/StorageAdapter.cs
namespace Greyfoundry.Identifold.Storage;

public sealed record ReferenceReservation(
    string MachineId,
    string Namespace,
    string Reference);

public sealed record ReferenceMapping(string MachineId, string Namespace);

public sealed record SequenceAllocationRequest(
    string MachineId,
    string Namespace,
    string ReferencePrefix,
    string? Scope,
    byte Width);

public interface IStorageAdapter
{
    ValueTask<bool> ReserveAsync(
        ReferenceReservation request,
        CancellationToken cancellationToken = default);

    ValueTask<ReferenceMapping?> ResolveAsync(
        string reference,
        string namespaceName,
        CancellationToken cancellationToken = default);

    ValueTask<ulong> AllocateAsync(
        SequenceAllocationRequest request,
        CancellationToken cancellationToken = default);
}

File: packages/swift/Sources/Identifold/Storage/StorageAdapter.swift
public struct ReferenceReservation: Equatable, Sendable {
  public let machineID: String
  public let namespace: String
  public let reference: String

  public init(machineID: String, namespace: String, reference: String) {
    self.machineID = machineID
    self.namespace = namespace
    self.reference = reference
  }
}

public struct ReferenceMapping: Equatable, Sendable {
  public let machineID: String
  public let namespace: String

  public init(machineID: String, namespace: String) {
    self.machineID = machineID
    self.namespace = namespace
  }
}

public struct SequenceAllocationRequest: Equatable, Sendable {
  public let machineID: String
  public let namespace: String
  public let referencePrefix: String
  public let scope: String?
  public let width: UInt8

  public init(
    machineID: String,
    namespace: String,
    referencePrefix: String,
    scope: String?,
    width: UInt8
  ) {
    self.machineID = machineID
    self.namespace = namespace
    self.referencePrefix = referencePrefix
    self.scope = scope
    self.width = width
  }
}

public protocol StorageAdapter: Sendable {
  func reserve(_ request: ReferenceReservation) async throws -> Bool

  func resolve(reference: String, namespace: String) async throws -> ReferenceMapping?

  func allocate(_ request: SequenceAllocationRequest) async throws -> UInt64
}
